test/compiled_metric_test.rb

```diff
@@ -0,0 +1,717 @@
+    )
+
+    metric.increment(value: 1)
+
+    datagram = @sink.datagrams.first
+    # Special characters should be converted to underscores
+    assert_equal("test.foo_bar_baz_qux", datagram.name)
+  end
+
+  def test_raises_on_unsupported_tag_type
+    assert_raises(ArgumentError) do
+      StatsD::Instrument::CompiledMetric::Counter.define(
+        name: "foo.bar",
+        tags: { invalid: Array },
+      )
+    end
+  end
+
+  def test_includes_default_tags_from_client
+    # Create a client with default tags
+    @sink = StatsD::Instrument::CaptureSink.new(parent: StatsD::Instrument::NullSink.new)
+    client = StatsD::Instrument::Client.new(
+      sink: @sink,
+      prefix: "test",
+      default_tags: ["env:production", "region:us-east"],
+      enable_aggregation: false,
+    )
+    StatsD.singleton_client = client
+
+    metric = StatsD::Instrument::CompiledMetric::Counter.define(
+      name: "foo.bar",
+      static_tags: { service: "web" },
+    )
+
+    metric.increment(value: 1)
+
+    datagram = @sink.datagrams.first
+    assert_equal("test.foo.bar", datagram.name)
+    # Should include default tags from client + static tags
+    assert_equal(["env:production", "region:us-east", "service:web"], datagram.tags.sort)
+  end
+
+  def test_includes_default_tags_with_no_prefix
+    # Create a client with default tags
+    @sink = StatsD::Instrument::CaptureSink.new(parent: StatsD::Instrument::NullSink.new)
+    client = StatsD::Instrument::Client.new(
+      sink: @sink,
+      prefix: "test",
+      default_tags: ["env:production", "region:us-east"],
+      enable_aggregation: false,
+    )
+    StatsD.singleton_client = client
+
+    metric = StatsD::Instrument::CompiledMetric::Counter.define(
+      name: "foo.bar",
+      static_tags: { service: "web" },
+      no_prefix: true,
+    )
+
+    metric.increment(value: 1)
+
+    datagram = @sink.datagrams.first
+    assert_equal("foo.bar", datagram.name) # No prefix
+    # Should include default tags even when no_prefix is true
+    assert_equal(["env:production", "region:us-east", "service:web"], datagram.tags.sort)
+  end
+
+  def test_custom_max_cache_size
+    metric = StatsD::Instrument::CompiledMetric::Counter.define(
+      name: "foo.bar",
+      tags: { shop_id: Integer },
+      max_cache_size: 2,
+    )
+
+    # Increment with 3 different tag combinations
+    metric.increment(shop_id: 1, value: 1)
+    metric.increment(shop_id: 2, value: 1)
+    metric.increment(shop_id: 3, value: 1)
+
+    # After 3 increments with max_cache_size of 2, cache should be cleared
+    # We can't easily test the internal cache state, but we can verify it doesn't crash
+    assert_equal(3, @sink.datagrams.size)
+  end
+
+  def test_cache_handles_hash_collisions
+    metric = StatsD::Instrument::CompiledMetric::Counter.define(
+      name: "foo.bar",
+      tags: { shop_id: Integer },
+    )
+
+    # Even if we get hash collisions, it should handle them correctly
```

Comment:
I doubt this will produce a collision, but you can stub any instance of `PrecompiledDatagram#hash` to return 1, and the test should still work. 

Do you think it would be possible to have the entire behaviour tested on the client irrespective of aggregation? I think the main change would be that you can't assert how many datagrams have been produced, but you can sum the counts